src/Npgsql/TypeHandlers/CompositeHandlers/MappedCompositeHandler.cs

```diff
@@ -0,0 +1,155 @@
+using System;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using System.Threading.Tasks;
+using Npgsql.BackendMessages;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeHandling;
+using Npgsql.TypeMapping;
+using NpgsqlTypes;
+
+namespace Npgsql.TypeHandlers.CompositeHandlers
+{
+    class MappedCompositeHandler<T> : NpgsqlTypeHandler<T>, IMappedCompositeHandler
+        where T : new()
+    {
+        static readonly Func<T> s_constructor = IsValueType<T>.Value
+            ? null
+            : Expression
+                .Lambda<Func<T>>(Expression.New(typeof(T)))
+                .Compile();
+
+        readonly CompositeMemberHandler<T>[] _members;
+
+        public Type CompositeType => typeof(T);
+
+        MappedCompositeHandler(PostgresCompositeType postgresType, CompositeMemberHandler<T>[] members)
+        {
+            PostgresType = postgresType;
+            _members = members;
+        }
+
+        public override async ValueTask<T> Read(NpgsqlReadBuffer buffer, int length, bool async, FieldDescription fieldDescription = null)
+        {
+            await buffer.Ensure(sizeof(int), async);
+
+            var fieldCount = buffer.ReadInt32();
+            if (fieldCount != _members.Length)
+                throw new InvalidOperationException($"pg_attributes contains {_members.Length} fields for type {PgDisplayName}, but {fieldCount} fields were received.");
+
+            if (IsValueType<T>.Value)
+            {
+                var composite = new ByReference<T>();
+                foreach (var member in _members)
+                    await member.Read(composite, buffer, async);
+
+                return composite.Value;
+            }
+            else
+            {
+                var composite = s_constructor();
```

Comment:
It's a good idea to eliminate instantiation with Activator, but... since T is already constrained to `new()`, why not just do `var composite = new T()`?

If we need to keep the code-generator for some reason, then a naming nit: we don't usually use naming such as `s_constructor`, how about `ClassConstructor` or `ClassInstantiator`?

Reply:
Because `new T()` actually is `Activator.CreateInstance<T>` because compile doesn't know an actual constructor which will be used. For structs it doesn't exist at all (:

Done, but i used the standard notation in .NET. I think about switching to it.

Reply:
> Because new T() actually is Activator.CreateInstance<T> because compile doesn't know an actual constructor which will be used.

I didn't know that - but I can see that you're right on sharplabs.io. Do you know of any blog/article that explains what it's like this? I'd have thought that as long as there's a `new()` type constraint, than an efficient `new T()` should be possible, without using Activator.

Reply:
See https://stackoverflow.com/a/367648/682105

Reply:
Wow, thanks for all those links @khellang :) Makes sense. Added a proposal on SO to have the JIT recognize `Activator.CreateInstance<T>()` and emit the right thing.

Reply:
A proposal on SO? By the way, I'm planning to open proposal in the Core FX repo in a few days and provide a non-intrinsic implementation.

Reply:
Maybe proposal was the wrong word, more like a thought :)

In any case, [see this comment](https://github.com/dotnet/coreclr/issues/7592#issuecomment-481699119)...